---
title: "packFinder"
author: "Jack Gisby"
date: "`r Sys.Date()`" 
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{packFinder}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(packFinder)
```

```{r paged.print=TRUE}
data("arabidopsisThalianaRefseq")

packMatches <- packSearch(
  Biostrings::DNAString("CACTACAA"),
  arabidopsisThalianaRefseq,
  mismatch = 0,
  elementLength = c(300, 3500),
  tsdLength = 3
)

head(packMatches)
```

```{r, eval=F, echo=T}
packMatches <- packClust(
  packMatches,
  arabidopsisThalianaRefseq,
  saveFolder = "data/",
  vSearchPath = "D:/vsearch-2.14.1-win-x86_64/vsearch.exe"
)
```

```{}
Rognes T, Flouri T, Nichols B, Quince C, Mahe F (2016)
VSEARCH: a versatile open source tool for metagenomics
PeerJ 4:e2584 doi: 10.7717/peerj.2584 https://doi.org/10.7717/peerj.2584

Compiled with support for gzip-compressed files, but the library was not found.
Compiled with support for bzip2-compressed files, but the library was not found.
vsearch v2.14.1_win_x86_64, 7.9GB RAM, 4 cores
https://github.com/torognes/vsearch

vsearch v2.14.1_win_x86_64, 7.9GB RAM, 4 cores
https://github.com/torognes/vsearch

Reading file data/packMatches.fasta 100%
9396 nt in 6 seqs, min 713, max 2463, avg 1566
Counting k-mers 100%
Clustering 100%
Sorting clusters 100%
Writing clusters 100%
Clusters: 5 Size min 1, max 2, avg 1.2
Singletons: 4, 66.7% of seqs, 80.0% of clusters
Sorting clusters by abundance 100%
```

```{r include=FALSE}

data(packMatches)
```

```{r}
print(packMatches)
```


```{r eval=FALSE, include=FALSE}
consensusSeqs <- tirClust(packMatches,
  arabidopsisThalianaRefseq,
  tirLength = 25,
  plotSavePath = "data/tirClusters.png"
)

head(consensusSeqs)
```

```{r Clear_File_Outputs, include=FALSE}
do.call(file.remove, list(list.files("data/", full.names = TRUE)))
```

