---
title: "packFinder"
author: "Jack Gisby"
date: "`r Sys.Date()`" 
output:
  rmarkdown::html_vignette:
    toc: TRUE
vignette: >
  %\VignetteIndexEntry{packFinder}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "# >",
  fig.width = 6,
  fig.height = 4
)
```


# Introduction
The goal of packFinder was to implement a simple tool for the prediction of potential Pack-TYPE elements. packFinder uses the following prior knowledge, provided by the user, to detect transposons:


1. Terminal Inverted Repeat (TIR) Base Sequence
2. Length of Terminal Site Duplication (TSD)
3. Length of the Transposon


These features, shown in __Figure 1__, provide enough information to detect autonomous and pack-TYPE elements. For a transposon to be predicted by packFinder its TSD sequences must be identical to each other, its forward TIR sequence must match the base sequence provided and its reverse TIR sequence must match its reverse complement.


![**Important structural features of Pack-TYPE transposons**](tirSeq.jpg){width=100%}



Transposons are therefore predicted by searching a given genome for these characteristics, and further analysis steps can reveal the nature of these elements - while the packFinder tool is sensitive for the detection of transposons, it does not discriminate between autonomous and Pack-TYPE elements. Autonomous elements will contain a transposase gene within the terminal inverted repeats and tend to be larger than their Pack-TYPE counterparts; pack-TYPE elements instead capture sections of host genomes. Following cluster analysis, BLAST can be used to discern which predicted elements are autonomous (transposase-containing) and with are true Pack-TYPE elements.


# Getting Started
Assuming the information specified in the introduction above is available, researchers may download packFinder and use the primary function - packSearch - to locate potential transposons in a given Genome. In addition to R packages, the command line tool "VSEARCH" must be installed prior to use of clustering and alignment functions.

## R Package Dependencies
In addition to CRAN package dependencies, that will be installed automatically upon downloading packFinder, some dependencies must be downloaded from Bioconductor. 

```{r}
source("https://bioconductor.org/biocLite.R")
biocLite()

biocLite(c("Biostrings", "GenomicRanges"))
```

Then, packFinder may be installed:

```{r eval=FALSE}
#devtools must be installed to download packFinder from github
#install.packages("devtools")

devtools::install_github("jackgisby/packFinder")
library("packFinder")
```
```{r setup}
knitr::opts_chunk$set(include = FALSE)
library("packFinder")
```

## Command Line Dependencies
While this allows R to run the packSearch pipeline within packFinder, VSEARCH must be installed for use of clustering and alignment functions within packSearch. Detailed installation instructions are available from the README file on the VSEARCH github (https://github.com/torognes/vsearch). The command line can be used to install VSEARCH on Linux and MacOS operating systems (using wget and tar) while VSEARCH can be downloaded and extracted for use on Windows systems.

# Searching for Potential Transposable Elements using packFinder
## Getting Data
Both DNA to be searched and the TIR query must be in Biostrings format. The TIR query should be coerced to a DNAString object while the DNA sequence, or set of sequences, should be a DNAStringSet. Biostrings provides methods for the conversion of various formats to DNAString objects, including from character vectors and fasta files. 
```{r eval=FALSE}
# Convert a character vector to a DNAString
Biostrings::DNAString("CATG")

# Convert a list of character vectors to a DNAStringSet
Biostrings::DNAString(c("CATG",
                        "GTAC"))

# Convert a FASTA file to a DNAStringSet
Biostrings::readDNAStringSet("path/to/fasta",
                             format="fasta")
```


## Using packSearch
In this example we will use a subset of chromosome 3 from the Arabidopsis thaliana reference sequence and the first 8 bases of the CACTA1 autonomous transposable element's TIR. The CACTA transposons have TSD sequences that are 3 bases long, and the Pack-CACTA elements tend to be between 300 to 3500 bases in width. Since the first 8 bases of the Pack-CACTA TIRs tend to be conserved between elements, the default allowable mismatch of 0 was used. 
```{r paged.print=TRUE}
data("arabidopsisThalianaRefseq")

packMatches <- packSearch(
  Biostrings::DNAString("CACTACAA"),
  arabidopsisThalianaRefseq,
  elementLength = c(300, 3500),
  tsdLength = 3
)

head(packMatches)
```

In this subset, we identified six potential Pack-TYPE elements, however at this stage it is unclear what genetic information can be found between the inverted TIR sequences __Figure 1__. These elements could contain transposase genes, making them autonomous elements, or may be Pack-TYPE elements that have captured parts of the Arabidopsis thaliana host genome. Additionally, of the Pack-TYPE elements detected, some may share sequences of related chromosomal origin between their TIRs. 

# Analysing Potential Transposable Elements using packFinder
## Clustering of Transposable Elements

In order to make downstream analysis more efficient and understand the relations between the identified elements, we can use VSEARCH to cluster predicted transposons. Here we run VSEARCH with the default parameters, meaning:
*An identity of 60% between two elements is required to form a cluster
*
```{r include=FALSE}
data(packMatches)
```
```{r, eval=F, echo=T}
packMatches <- packClust(
  packMatches,
  arabidopsisThalianaRefseq,
  saveFolder = "data/",
  vSearchPath = "D:/vsearch-2.14.1-win-x86_64/vsearch.exe"
)
```
```{}
Rognes T, Flouri T, Nichols B, Quince C, Mahe F (2016)
VSEARCH: a versatile open source tool for metagenomics
PeerJ 4:e2584 doi: 10.7717/peerj.2584 https://doi.org/10.7717/peerj.2584

Compiled with support for gzip-compressed files, but the library was not found.
Compiled with support for bzip2-compressed files, but the library was not found.
vsearch v2.14.1_win_x86_64, 7.9GB RAM, 4 cores
https://github.com/torognes/vsearch

vsearch v2.14.1_win_x86_64, 7.9GB RAM, 4 cores
https://github.com/torognes/vsearch

Reading file data/packMatches.fasta 100%
9396 nt in 6 seqs, min 713, max 2463, avg 1566
Counting k-mers 100%
Clustering 100%
Sorting clusters 100%
Writing clusters 100%
Clusters: 5 Size min 1, max 2, avg 1.2
Singletons: 4, 66.7% of seqs, 80.0% of clusters
Sorting clusters by abundance 100%

  seqnames   start     end width strand TSD cluster
1     Chr3  100830  102347  1518      + TGT       3
2     Chr3 1068802 1069514   713      + ATA       4
3     Chr3 2807747 2809454  1708      + GGT       2
4     Chr3 3487540 3488267   728      + ATA       4
5     Chr3 3582297 3584562  2266      + ATA       1
6     Chr3 3738747 3741209  2463      + TTG       0
```

## Clustering of TIR Sequences
```{r echo=TRUE}
consensusSeqs <- tirClust(packMatches,
  arabidopsisThalianaRefseq,
  tirLength = 25,
  plotSavePath = "data/tirClusters.png"
)

head(consensusSeqs)
```

## Alignment of Transposable Elements

# Data Formats and Conversion
## packSearch Dataframe

## GRanges

## FASTA

```{r Clear_File_Outputs, include=FALSE}
do.call(file.remove, list(list.files("data/", full.names = TRUE)))
```