#' @title
#' Global Alignment with VSEARCH
#'
#' @description
#' A global pairwise alignment of pack-TYPE elements by sequence similarity.
#' It may be useful to run \code{\link{packClust}} to identify groups of
#' similar transposable elements, before generating alignments of each group.
#'
#' @param packMatches
#' A dataframe of potential Pack-TYPE transposable elements. Will be saved as a
#' FASTA file for VSEARCH.
#'
#' @param Genome
#' A DNAStringSet object containing sequences referred to in \code{packMatches}
#' (the object originally used to predict the transposons
#' \code{\link{packSearch}}).
#'
#' @param identity
#' The sequence identity of two transposable elements in \code{packMatches}
#' required to be grouped into a cluster.
#'
#' @param threads
#' The number of threads to be used by VSEARCH.
#'
#' @param identityDefinition
#' The pairwise identity definition used by VSEARCH. Defaults to 1, the BLAST
#' definition.
#'
#' @param saveFolder
#' The folder to save saveFolder files (uc, blast6out, FASTA)
#'
#' @param vSearchPath
#' The location of the VSEARCH executable file.
#'
#' @note
#' In order to align sequences using VSEARCH, the executable file must first
#' be installed.
#'
#' @author
#' Jack Gisby
#'
#' @return
#' Saves alignment information, including a \code{uc}, \code{blast6out} and a
#' pairwise alignment \code{fasta} file, to the specified location. Returns the
#' uc summary file generated by the alignment.
#'
#' @references
#' VSEARCH may be downloaded from \url{https://github.com/torognes/vsearch}
#' along with a manual documenting the program's parameters. See
#' \url{https://www.ncbi.nlm.nih.gov/pubmed/27781170} for further information.
#'
#' @seealso
#' code{\link{tirClust}}, code{\link{packClust}}, code{\link{readBlast6Out}},
#' code{\link{readUc}}
#'
#' @export

packAlign <- function(packMatches,
                      Genome,
                      identity = 0,
                      threads = 1,
                      identityDefinition = 1,
                      saveFolder,
                      vSearchPath = "path/to/vsearch/vsearch-2.14.1-win-x86_64/vsearch.exe") {
  if (is.null(saveFolder)) {
    saveFolder <- getwd()
  }

  if (parallel::detectCores() < threads) {
    stop("There are not ", threads, " cores available")
  }

  if (identity > 1 | identity < 0) {
    stop("Identity must be of type integer or double, and have a value between 0 and 1")
  }

  if (!is.integer(identityDefinition)) {
    stop("Argument 'identityDefinition' must be an integer between 0 and 4.")
  } else if (identityDefinition > 4 | identityDefinition < 0) {
    stop("Argument 'identityDefinition' must be an integer between 0 and 4.")
  }

  if (system2(vSearchPath, "--v") != 0) {
    message("VSEARCH cannot be found at", vSearchPath, ". Please ensure that
            the VSEARCH executable file is installed at the correct location
            and that the full file path is specified.")
  }

  packMatchesFile <- paste0(saveFolder, "/packMatches.fasta")
  packMatchesSet <- getPackSeqs(packMatches, Genome, output = "DNAStringSet")
  packMatchesSet@ranges@NAMES <- as.character(rownames(packMatches))
  Biostrings::writeXStringSet(packMatchesSet, packMatchesFile)

  system2(
    command = vSearchPath,
    args = paste0(
      "--allpairs_global ", packMatchesFile, " \ ",
      "--id ", identity, " \ ",
      "--output_no_hits \ ",
      "--iddef ", identityDefinition, " \ ",
      "--threads ", threads, " \ ",
      "--uc ", file.path(saveFolder, paste0("vSearchPairwiseAlignment", ".uc")), " \ ",
      "--fastapairs ", file.path(saveFolder, paste0("alignmentPairs.fasta")), " \ ",
      "--blast6out ", file.path(saveFolder, paste0("vSearchPairwiseAlignment", ".blast6out")), " \ "
    )
  )

  return(readUc(file.path(saveFolder, paste0("vSearchPairwiseAlignment", ".uc")), output = "alignment"))
}
